Simplify update script, add missing use-ffmpeg-release, support custom commits

avih opened the pull request.

avih commented:
Forced push the last commit, just removed accidental extra empty line.

kevmitch commented on update:
```diff
@@ -59,19 +65,35 @@ do_fixedref()
+
+# fallback targets: release/master if no config file
+checkout_ffmpeg=master
+#checkout_fribidi=release
```
Let's just remove any mention of fribidi.

avih replied:
Actually, as someone who can read (some) RTL text, I can attest it's really important for RTL text, and without it RTL subs are displayed wrongly.

Unless you you feel strongly about removing it, I think it'd be useful to keep it.

wm4 replied:
Fribidi isn't optional in libass, but most time the system version is fine.

avih replied:
Gotcha. Sure, I'll remove it then.

avih replied:
> Gotcha. Sure, I'll remove it then.

Actually, I have plans to keep supporting fribidi in a formal way but keeping it disabled by default in mpv-build, following wm4'a comment.

So I'll leave it commented out for now, and if my plan doesn't pan out then we can remove the comments later.

kevmitch replied:
Where is a non-system fribidi needed? Is it more important than other dependencies that are not included in mpv-build? My understanding is that this was added years ago with little explanation (4d6f1256116aa785992e2ed1df78c1d9932ceb3b) for someone's special case, but actually building it has been disabled (b38387044631aa433ae1612efbe9b1ed1e264b51) for most of that time with no one complaining.

avih replied:
> Where is a non-system fribidi needed?

I don't know. I based it on wm4's comment which I interpreted as there are system where the global fribidi is not good enough.

> Is it more important than other dependencies that are not included in mpv-build?

There are a lot of them and I didn't line them up with priorities, but based on the fact that fribidi was there and others were not, seems to suggest that it might. But that's lots of ifs and buts.

However, it's definitely easier to support because we already have the config and build scripts for it, and we don't for others.

In general, my plan is to structure it in a way that it's easy to add more. The framework for config/build is already there to be extended relatively easily, but the update script is the only part which is a bit hardcoded towards a specific set of libs.

So moving the use-foo-bar scripts into a single config file would be one step in making it easy to extend without inflating the scripts count, and the update script hopefully can be maded similarly simple to extend with new libs.

Fribidi is just one which is both relatively easy to support and can show off the capabilities of having something formally supported yet disabled by default (where disabled here means to use the system libs).

kevmitch commented on update:
```diff
@@ -59,19 +65,35 @@ do_fixedref()
-#checkout_fribidi=do_releasetag
-checkout_libass=do_gitmaster
-checkout_mpv=do_gitmaster
+# args: $1: project name, $2: release/master [,$3: non-empty to ignore the config file]
```
why not
* $2 non-empty -> use that
* $2 empty -> use config
    * config empty -> use master

The only downside would be that you can't then set different defaults on different packages, but I don't think we really care about that now since it's all master. If that ever changes, I think $2 should be the default and $3 should be empty or override value.

avih replied:
> why not
> 
>     $2 non-empty -> use that
>     $2 empty -> use config
>         config empty -> use master

Because this function is called with one of two use cases (which I didn't modify):
- Try the config file, and if it doesn't exist then use the provided value.
- Just use the provided value.

And the API supports exactly these two cases, while your suggestion does not.

Also, `master` as fallback happens to be the current _choice_ as default value for all projects, but IMO it's not something which should be taken for granted by disallowing any other default values, which your suggestion would do.

In general, I think your approach would make the code more implicit in a way which is less easy to modify further. My goal was not to reduce the number of chars at the script, but rather make it easier to modify further.

kevmitch replied:
Ok, I just felt kind of weird about the "-" argument.

avih replied:
> Ok, I just felt kind of weird about the "-" argument.

I completely agree about this. One of previous versions of this API had $2 as a config file name to try first, but ultimately I decided that the config file is always derived from the name, so it's more correct to indicate whether or not to use a config file rather than providing or not providing a redundant config file name.

It's not too bad, but I agree it's not amazing either. But it's better than before, which was the main goal ;)

kevmitch commented on update:
```diff
@@ -59,19 +65,35 @@ do_fixedref()
+# args: $1: project name, $2: release/master [,$3: non-empty to ignore the config file]
+checkout()
+{
+    local arg="$2"
```
`branch` would be more descriptive than arg.

avih replied:
It's not always a branch either, because the next commit adds support for a custom commit/tag/branch. Maybe `version` or `checkout` would have been a more appropriate name. But I also didn't want to modify the config file names branch-$project . So yeah, I'll make it `branch`.

> I hate to suggest adding more scripts, but what if someone wants a specific libass version? I think the reason it currently doesn't have such scripts is only to discourage use of its releases which aren't kept up to date. Maybe a use-libass-master, use-libass-custom are in order?

I feel the same. In general, this whole `update` script is just supposed to make it easier for people to pull and possibly checkout a specific revision of the projects. The only "smart" thing it knows to do otherwise is to map "release" to the latest release tag.

All those scripts are just supposed to make it easier to set a value of release/master/custom per project.

How about this: I'll add 3 more libass scripts for symmetry (or just 2 like you suggested, I don't mind at all), and then do a followup to unify all those 8/9 scripts, together with the additional projects options, to one config file which could have comments in it etc?

This single config would be a one-stop-shop for anyone who wants to customize it. Makes sense?

kevmitch replied:
That sounds fine, though I think I'd lean towards not having a libass master script, if nothing else than to reduce the sheer number of scripts unless you're suggesting somehow merging them too. There is something to be said about `ls` basically providing "documentation", but there's a point where more scripts are more inundating than helpful.

avih replied:
> unless you're suggesting somehow merging them too

That's what I'm suggesting. One config file to replace all those use-foo-bar scripts, together with the options per project config values. All the config in one place, with a default config file which can be copied and modified by users, or some such.

kevmitch replied:
Oh, yeah, sure that works. Much simpler. I guess it's not too much to ask to edit a file if you're that particular about your versions.

avih commented:
Updated last commit to use the var name `branch` and added two scripts: use-libass-{master|custom} .

Overall, this PR adds adds support to use custom tag/branch/hash as a version to build for mpv/ffmpeg/libass via the scripts `use-{mpv|ffmpeg|libass}-custom <a thing to checkout>`.

Let's leave the fribidi commented out for now, and if we can't handle it nicely soon then we can remove it.
